fix: add validation check and enable button only valid status in import and run page

mihilt opened the pull request:
#1387

To solve this problem, I created a variable called `importButtonDisabled`.
Checking length of textfield's input value through `@change` property and it can handle `importButtonDisabled` value properly.
The button dynamically changes the property called `disabled` according to the variable.

mihilt commented:
This comment is about the second commit.
I checked in the first commit that the function what i made works only when i move focus off the textfield.
So I changed `@change` to `@keyup`.

mihilt commented:
I am sorry, I just found that i need to fix three more buttons.

I will fix the remaining buttons in the same way as I committed before.
But I don't know that this way is right. Can someone give me an advice?

lizable commented:
> I am sorry, I just found that I need to fix three more buttons.
> 
> I will fix the remaining buttons in the same way as I committed before. But I don't know that this way is right. Can someone give me an advice?

Hi, @mihilt! 
First of all, Thanks for making your effort on this open-source project!
Speaking of your request, I think it's better to describe the issue with several tasks so that you can easily understand what to do.

In this issue, there are two tasks to achieve.
1. Add validation logic(including showing up message with snackbar UI) in two parts:
   - [ ] input field in Import Notebook panel
   - [ ] input field in Create Notebook button panel
2. Enable each button on "Import & Run" page only when the corresponding input value is valid.

FWIW, I'm not sure about adding a property for each button. Could we just set it as disabled at the first render and change it when the input field is valid?

mihilt commented:
Thank you for advicing @lizable.
I'm sorry. I should have read the issue in more detail. I thought the thing about snackbar was optional.

BTW, I have two questions.

> - [ ] input field in Create Notebook button panel

I think the input field in **Create Notebook Button** panel already has its validation logic.
Cuz I saw snackbar with message when I clicked button with empty input field in this panel.

And I didn't found any snackbar with message when I clicked button with empty input field in **Import Github Repository** panel.
So I think I should add validation logic in **Import Github Repository** panel instead of **Create Notebook Button** panel.
Is it right?

> FWIW, I'm not sure about adding a property for each button. Could we just set it as disabled at the first render and change it when the input field is valid?
> 

I'm not sure, but In this case, is there have a way to handle if the valid input value is changed back to an invalid input value?
(In more detail, if it's textfield's value goes from invalid to valid and invalid again. In this situation, I think the button can be clicked even if there has a problem with the input value.)
I think each button needs their status value. So it needs its own property with each separated variable.

I think I bothered you a little bit with a lot of questions even if this issue has **easy** and **good first issue** label.
It's because I'm a beginner contributor in this repo. I'm sorry.

lizable commented:
> 



> Thank you for advicing jihyunkang (@lizable). I'm sorry. I should have read the issue in more detail. I thought the thing about snackbar was optional.
> 
> BTW, I have two questions.
> 
> > * [ ]  input field in Create Notebook button panel
> 
> I think the input field in **Create Notebook Button** panel already has its validation logic. Cuz I saw snackbar with message when I clicked button with empty input field in this panel.
> 
> And I didn't found any snackbar with message when I clicked button with empty input field in **Import Github Repository** panel. So I think I should add validation logic in **Import Github Repository** panel instead of **Create Notebook Button** panel. Is it right?
> 
> > FWIW, I'm not sure about adding a property for each button. Could we just set it as disabled at the first render and change it when the input field is valid?
> 
> I'm not sure, but In this case, is there have a way to handle if the valid input value is changed back to an invalid input value? (In more detail, if it's textfield's value goes from invalid to valid and invalid again. In this situation, I think the button can be clicked even if there has a problem with the input value.) I think each button needs their status value. So it needs its own property with each separated variable.
> 
> I think I bothered you a little bit with a lot of questions even if this issue has **easy** and **good first issue** label. It's because I'm a beginner contributor in this repo. I'm sorry.

You don't need to feel sorry about it. If you have anything uncertain while handling the issue, It's okay to ask me anytime.
My answer to the first question is that we need to apply validation logic not only for the empty string but also the format.
As long as I know, only opened URL (accessible in public) for notebook file is available for making an instant session and a badge button. Therefore you need to apply validation logic that checks whether the user input is empty and also whether the user input is valid URL(e.g. starting from `https://`). This code block below may help you to figure out the regex for valid URL.
https://github.com/lablup/backend.ai-webui/blob/77f284acdb6e7b88d70eee772d608bbfef9ee39b/src/components/backend-ai-registry-list.ts#L659

And I think we don't need to care about frequent state changing of input field, (including as you mentioned above, invalid->valid->invalid) when we replace `keyup` event handler to `change`. Unlikely to `keyup` event, `change` event triggered when user input field is blurred, which refers that the user input has been finished.
Therefore I suggest using `change` event rather than `keyup` event, like most input field in backend.ai-webui do.

lizable reviewed (changes requested):
Even if the URL is valid, input field keep saying it's invalid.
Here's the URL address that I used:
- jupyter notebook file shared in public
https://github.com/hephaex/deeplearning-note/blob/master/08_dimensionality_reduction.ipynb
- GitHub address shared in public
https://github.com/rickiepark/handson-ml2.git



### Screenshots
![not-working mov](https://user-images.githubusercontent.com/46954439/184792092-e854039e-4b7e-412a-a730-accbb310034d.gif)

mihilt commented:
Thank you for telling me. I changed regex.

https://user-images.githubusercontent.com/50368758/184861995-76c0e18b-f93f-4907-b322-4974f29ae939.mp4

inureyes commented:
@mihilt mwc-textfield has custom validation methods such as `checkValidity()`, `reportValidity()` and `setCustomValidity()` Please refer the link https://www.npmjs.com/package/@material/mwc-textfield and revise the implementation to use official validation logic. :)

You already wrote the validation code therefore I think small modification to bind the validation logic with proper way will be necessary.

@lizable I mentioned you to explain the proper validation logic too.

lizable commented:
> @mihilt mwc-textfield has custom validation methods such as `checkValidity()`, `reportValidity()` and `setCustomValidity()` Please refer the link https://www.npmjs.com/package/@material/mwc-textfield and revise the implementation to use official validation logic. :)
> 
> You already wrote the validation code therefore I think small modification to bind the validation logic with proper way will be necessary.
> 
> jihyunkang (@lizable) I mentioned you to explain the proper validation logic too.

I agree. We already have custom validation logic of `mwc-textfield`, so you may refer to the code below and apply it to this PR.
https://github.com/lablup/backend.ai-webui/blob/64048cefb6aa66a6b8c68ade49f5ba29d9fef54a/src/components/backend-ai-signup.ts#L334-L364

mihilt commented:
@inureyes @lizable
Thank you for kind explanation. I think I changed it as you wanted.
I deleted validation functions that existed before and used mwc-textfield's method and properties.

https://user-images.githubusercontent.com/50368758/185409481-c4a6557a-7612-46f9-ad83-7576ab52a7f8.mp4

lizable commented on src/components/backend-ai-import-view.ts:
```diff
@@ -581,6 +581,18 @@ export default class BackendAIImport extends BackendAIPage {
+    } else {
+      this.notification.text = _text('import.WrongURLType');
+      button.setAttribute('disabled', 'true');
+      this.importMessage = this.notification.text;
```
I guess we should split the importMessage to import notebook and import GitHub repo, since when either the value of the input field of Import Notebook or Import From GitHub is invalid, the error message is shown in both areas.
Here are the screenshots that refer to the issue that I mentioned above.

### Screnshots
![GitHub_Repo mov](https://user-images.githubusercontent.com/46954439/185544627-5f1696e5-57f6-4382-97bd-293e71a75ee1.gif)
<img width="673" alt="ImportNotebookRepo" src="https://user-images.githubusercontent.com/46954439/185544651-b74117dc-2417-42f0-8396-9358d205f849.png">

lizable reviewed (changes requested):
I confirmed that validation logic is okay, but there's one minor thing to be better fixed.
Please refer to the comment below. except that, all is good!

mihilt commented:
It is just my curious, but I wanted to know about never used lit's @property status value.

I found that some values 'condition', 'resourcePolicy', '_helpDescriptionTitle', '_helpDescriptionIcon', 'storageProxyInfo' are never used in backend-ai-import-view.ts file.

I think it is hard to notice because ESLint can't check it.
Is it just useless value? Or not used in the component, but necessary?
If it is a useless value, can I delete it all?

lizable commented:
> It is just my curious, but I wanted to know about never used lit's @Property status value.
> 
> I found that some values 'condition', 'resourcePolicy', '_helpDescriptionTitle', '_helpDescriptionIcon', 'storageProxyInfo' are never used in backend-ai-import-view.ts file.
> 
> I think it is hard to notice because ESLint can't check it. Is it just useless value? Or not used in the component, but necessary? If it is a useless value, can I delete it all?

I confirmed that those properties described above seem to be legacy, but some of them should be left as is because they were used for the backbone component. Therefore let's remove only two properties, `resourcePolicy` and `storageProxyInfo` for now.

lizable reviewed (approved):
LGTM.
Well done, @mihilt! 👍

Btw, I just renamed the PR title to be a more precise description of what we want to achieve through this issue and PR.
FYI, we already have [the guide(Writing change logs)](https://github.com/lablup/backend.ai/tree/main/changes) for writing a good commit message, descriptions, etc, written by @achimnol.
Please refer to the guide, I'm sure that It will help you and other contributors a lot!
